File: usecases/src/test/java/com/zeyad/usecases/api/DataServiceTest.java
package com.zeyad.usecases.api;

import com.zeyad.usecases.TestRealmModel;
import com.zeyad.usecases.db.RealmQueryProvider;
import com.zeyad.usecases.requests.FileIORequest;
import com.zeyad.usecases.requests.GetRequest;
import com.zeyad.usecases.requests.PostRequest;
import com.zeyad.usecases.stores.CloudStore;
import com.zeyad.usecases.stores.DataStoreFactory;
import com.zeyad.usecases.stores.DiskStore;
import com.zeyad.usecases.stores.MemoryStore;

import org.json.JSONArray;
import org.json.JSONObject;
import org.junit.Before;
import org.junit.Test;

import java.io.File;
import java.util.Collections;
import java.util.HashMap;

import io.reactivex.Flowable;
import io.reactivex.Scheduler;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;

import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyBoolean;
import static org.mockito.Matchers.anyMap;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * @author by ZIaDo on 5/9/17.
 */
public class DataServiceTest {

    private DataService dataService;
    private DataStoreFactory dataStoreFactory;
    private GetRequest getRequest;
    private PostRequest postRequest;
    private Flowable flowable;

    @Before
    public void setUp() throws Exception {
        flowable = Flowable.just(true);
        postRequest = new PostRequest.Builder(Object.class, false).build();
        getRequest = new GetRequest.Builder(TestRealmModel.class, false)
                .url("")
                .id(37, "id", int.class)
                .cache("id")
                .build();
        dataStoreFactory = mock(DataStoreFactory.class);
        when(dataStoreFactory.dynamically(anyString(), any(Class.class)))
                .thenReturn(mock(CloudStore.class));
        when(dataStoreFactory.disk(any())).thenReturn(mock(DiskStore.class));
        when(dataStoreFactory.cloud(any())).thenReturn(mock(CloudStore.class));
        when(dataStoreFactory.memory()).thenReturn(mock(MemoryStore.class));
        dataService =
                new DataService(
                        dataStoreFactory, AndroidSchedulers.mainThread(), mock(Scheduler.class));
        com.zeyad.usecases.Config.setWithCache(false);
        com.zeyad.usecases.Config.setWithSQLite(true);
    }

    @Test
    public void getList() throws Exception {
        when(dataStoreFactory
                .dynamically(anyString(), any(Class.class))
                .dynamicGetList(anyString(), anyString(), any(Class.class), anyBoolean(), anyBoolean()))
                .thenReturn(Flowable.just(Collections.EMPTY_LIST));

        dataService.getList(getRequest);

        verify(dataStoreFactory.dynamically(anyString(), any(Class.class)), times(1))
                .dynamicGetList(anyString(), anyString(), any(Class.class), anyBoolean(), anyBoolean());
    }

    @Test
    public void getObject() throws Exception {
        when(dataStoreFactory
                .dynamically(anyString(), any(Class.class))
                .dynamicGetObject(
                        anyString(),
                        anyString(),
                        any(),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(flowable);

        dataService.getObject(getRequest);

        verify(dataStoreFactory.dynamically(anyString(), any(Class.class)), times(1))
                .dynamicGetObject(
                        anyString(),
                        anyString(),
                        any(),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean());
    }

    @Test
    public void getListOffLineFirst() throws Exception {
        when(dataStoreFactory
                .cloud(Object.class)
                .dynamicGetList(anyString(), anyString(), any(Class.class), anyBoolean(), anyBoolean()))
                .thenReturn(Flowable.just(Collections.EMPTY_LIST));
        when(dataStoreFactory
                .disk(Object.class)
                .dynamicGetList(anyString(), anyString(), any(Class.class), anyBoolean(), anyBoolean()))
                .thenReturn(Flowable.just(Collections.EMPTY_LIST));
        when(dataStoreFactory
                .memory()
                .getAllItems(any(Class.class)))
                .thenReturn(Single.just(Collections.singletonList(true)));

        dataService.getListOffLineFirst(getRequest);

        verify(dataStoreFactory.cloud(Object.class), times(1))
                .dynamicGetList(anyString(), anyString(), any(Class.class), anyBoolean(), anyBoolean());
        verify(dataStoreFactory.disk(Object.class), times(1))
                .dynamicGetList(anyString(), anyString(), any(Class.class), anyBoolean(), anyBoolean());
    }

    @Test
    public void getObjectOffLineFirst() throws Exception {
        when(dataStoreFactory.cloud(Object.class)
                .dynamicGetObject(
                        anyString(),
                        anyString(),
                        any(),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(flowable);
        when(dataStoreFactory
                .disk(Object.class)
                .dynamicGetObject(
                        anyString(),
                        anyString(),
                        any(),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(flowable);

        when(dataStoreFactory
                .memory()
                .getItem(anyString(), any(Class.class)))
                .thenReturn(Single.just(true));

        dataService.getObjectOffLineFirst(getRequest);

        verify(dataStoreFactory.cloud(Object.class), times(1))
                .dynamicGetObject(
                        anyString(),
                        anyString(),
                        any(),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean());
        verify(dataStoreFactory.disk(Object.class), times(1))
                .dynamicGetObject(
                        anyString(),
                        anyString(),
                        any(),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean());
    }

    @Test
    public void patchObject() throws Exception {
        when(dataStoreFactory.dynamically(anyString(), any(Class.class))
                .dynamicPatchObject(
                        anyString(),
                        anyString(),
                        any(Class.class),
                        any(JSONObject.class),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(flowable);

        dataService.patchObject(postRequest);

        verify(dataStoreFactory.dynamically(anyString(), any(Class.class)), times(1))
                .dynamicPatchObject(
                        anyString(),
                        anyString(),
                        any(Class.class),
                        any(JSONObject.class),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean(),
                        anyBoolean());
    }

    @Test
    public void postObject() throws Exception {
        when(dataStoreFactory
                .dynamically(anyString(), any(Class.class))
                .dynamicPostObject(
                        anyString(),
                        anyString(),
                        any(Class.class),
                        any(JSONObject.class),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(flowable);

        dataService.postObject(postRequest);

        verify(dataStoreFactory.dynamically(anyString(), any(Class.class)), times(1))
                .dynamicPostObject(
                        anyString(),
                        anyString(),
                        any(Class.class),
                        any(JSONObject.class),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean(),
                        anyBoolean());
    }

    @Test
    public void postList() throws Exception {
        when(dataStoreFactory
                .dynamically(anyString(), any(Class.class))
                .dynamicPostList(
                        anyString(),
                        anyString(),
                        any(Class.class),
                        any(JSONArray.class),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(flowable);

        dataService.postList(postRequest);

        verify(dataStoreFactory.dynamically(anyString(), any(Class.class)), times(1))
                .dynamicPostList(
                        anyString(),
                        anyString(),
                        any(Class.class),
                        any(JSONArray.class),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean(),
                        anyBoolean());
    }

    @Test
    public void putObject() throws Exception {
        when(dataStoreFactory
                .dynamically(anyString(), any(Class.class))
                .dynamicPutObject(
                        anyString(),
                        anyString(),
                        any(Class.class),
                        any(JSONObject.class),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(flowable);

        dataService.putObject(postRequest);

        verify(dataStoreFactory.dynamically(anyString(), any(Class.class)), times(1))
                .dynamicPutObject(
                        anyString(),
                        anyString(),
                        any(Class.class),
                        any(JSONObject.class),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean(),
                        anyBoolean());
    }

    @Test
    public void putList() throws Exception {
        when(dataStoreFactory
                .dynamically(anyString(), any(Class.class))
                .dynamicPutList(
                        anyString(),
                        anyString(),
                        any(Class.class),
                        any(JSONArray.class),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(flowable);

        dataService.putList(postRequest);

        verify(dataStoreFactory.dynamically(anyString(), any(Class.class)), times(1))
                .dynamicPutList(
                        anyString(),
                        anyString(),
                        any(Class.class),
                        any(JSONArray.class),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean(),
                        anyBoolean());
    }

    @Test
    public void deleteItemById() throws Exception {
        when(dataStoreFactory
                .dynamically(anyString(), any(Class.class))
                .dynamicDeleteCollection(
                        anyString(),
                        anyString(),
                        any(Class.class),
                        any(JSONArray.class),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(flowable);

        dataService.deleteItemById(postRequest);

        verify(dataStoreFactory.dynamically(anyString(), any(Class.class)), times(1))
                .dynamicDeleteCollection(
                        anyString(),
                        anyString(),
                        any(Class.class),
                        any(JSONArray.class),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean(),
                        anyBoolean());
    }

    @Test
    public void deleteCollection() throws Exception {
        when(dataStoreFactory
                .dynamically(anyString(), any(Class.class))
                .dynamicDeleteCollection(
                        anyString(),
                        anyString(),
                        any(Class.class),
                        any(JSONArray.class),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(flowable);

        dataService.deleteCollectionByIds(postRequest);

        verify(dataStoreFactory.dynamically(anyString(), any(Class.class)), times(1))
                .dynamicDeleteCollection(
                        anyString(),
                        anyString(),
                        any(Class.class),
                        any(JSONArray.class),
                        any(Class.class),
                        any(Class.class),
                        anyBoolean(),
                        anyBoolean(),
                        anyBoolean());
    }

    @Test
    public void deleteAll() throws Exception {
        when(dataStoreFactory.disk(Object.class).dynamicDeleteAll(any(Class.class)))
                .thenReturn(Single.just(true));

        dataService.deleteAll(postRequest);

        verify(dataStoreFactory.disk(Object.class), times(1)).dynamicDeleteAll(any(Class.class));
    }

    @Test
    public void queryDisk() throws Exception {
        when(dataStoreFactory.disk(Object.class).queryDisk(any(RealmQueryProvider.class)))
                .thenReturn(flowable);

        dataService.queryDisk(realm -> realm.where(TestRealmModel.class));

        verify(dataStoreFactory.disk(Object.class), times(1))
                .queryDisk(any(RealmQueryProvider.class));
    }

    @Test
    public void uploadFile() throws Exception {
        when(dataStoreFactory
                .cloud(Object.class)
                .dynamicUploadFile(
                        anyString(),
                        any(File.class),
                        anyString(),
                        (HashMap<String, Object>) anyMap(),
                        anyBoolean(),
                        anyBoolean(),
                        anyBoolean(),
                        any(Class.class)))
                .thenReturn(flowable);

        dataService.uploadFile(
                new FileIORequest.Builder("", new File("")).dataClass(Object.class).build());

        verify(dataStoreFactory.cloud(Object.class), times(1))
                .dynamicUploadFile(
                        anyString(),
                        any(File.class),
                        anyString(),
                        (HashMap<String, Object>) anyMap(),
                        anyBoolean(),
                        anyBoolean(),
                        anyBoolean(),
                        any(Class.class));
    }

    @Test
    public void downloadFile() throws Exception {
        when(dataStoreFactory
                .cloud(Object.class)
                .dynamicDownloadFile(
                        anyString(),
                        any(File.class),
                        anyBoolean(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(flowable);

        dataService.downloadFile(
                new FileIORequest.Builder("", new File("")).dataClass(Object.class).build());

        verify(dataStoreFactory.cloud(Object.class), times(1))
                .dynamicDownloadFile(
                        anyString(), any(File.class), anyBoolean(), anyBoolean(), anyBoolean());
    }
}

File: usecases/src/test/java/com/zeyad/usecases/api/DataServiceTest.kt
package com.zeyad.usecases.api

import android.support.test.rule.BuildConfig
import com.zeyad.usecases.TestModel
import com.zeyad.usecases.anyObject
import com.zeyad.usecases.requests.FileIORequest
import com.zeyad.usecases.requests.GetRequest
import com.zeyad.usecases.requests.PostRequest
import com.zeyad.usecases.stores.*
import io.reactivex.Flowable
import io.reactivex.Scheduler
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Matchers.*
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import java.io.File
import java.util.*

/**
 * @author by ZIaDo on 5/9/17.
 */
@RunWith(RobolectricTestRunner::class)
@Config(constants = BuildConfig::class, sdk = [25])
class DataServiceTest {

    private lateinit var dataService: DataService
    private lateinit var dataStoreFactory: DataStoreFactory
    private lateinit var getRequest: GetRequest
    private lateinit var postRequest: PostRequest
    private lateinit var flowable: Flowable<Any>
    private lateinit var single: Single<Any>
    private lateinit var fileFlowable: Single<File>

    @Before
    @Throws(Exception::class)
    fun setUp() {
        flowable = Flowable.just(true)
        single = Single.just(true)
        fileFlowable = Single.just(File(""))
        postRequest = PostRequest.Builder(TestModel::class.java, false).payLoad(Any()).build()
        getRequest = GetRequest.Builder(TestModel::class.java, false)
                .url("")
                .id(37, "id")
                .cache("id")
                .build()
        dataStoreFactory = mock(DataStoreFactory::class.java)
        `when`<DataStore>(dataStoreFactory.dynamically(anyString(), anyObject()))
                .thenReturn(mock(CloudStore::class.java))
        `when`<DataStore>(dataStoreFactory.disk(anyObject())).thenReturn(mock(DiskStore::class.java))
        `when`<DataStore>(dataStoreFactory.cloud(anyObject())).thenReturn(mock(CloudStore::class.java))
        `when`(dataStoreFactory.memory()).thenReturn(mock(MemoryStore::class.java))
        dataService = DataService(
                dataStoreFactory, AndroidSchedulers.mainThread(), mock(Scheduler::class.java))
        com.zeyad.usecases.Config.withCache = false
        com.zeyad.usecases.Config.withSQLite = true
    }

    @Test
    @Throws(Exception::class)
    fun getList() {
        `when`(dataStoreFactory
                .dynamically(anyString(), anyObject())
                .dynamicGetList<Any>(anyString(),
                        anyString(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(Flowable.just(mutableListOf()))

        dataService.getList<TestModel>(getRequest)

        verify<DataStore>(dataStoreFactory.dynamically(anyString(), anyObject()), times(1))
                .dynamicGetList<Any>(anyString(),
                        anyString(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean())
    }

    @Test
    @Throws(Exception::class)
    fun getObject() {
        `when`(dataStoreFactory
                .dynamically(anyString(), anyObject())
                .dynamicGetObject<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(flowable)

        dataService.getObject<TestModel>(getRequest)

        verify<DataStore>(dataStoreFactory.dynamically(anyString(), anyObject()), times(1))
                .dynamicGetObject<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean())
    }

    @Test
    @Throws(Exception::class)
    fun getListOffLineFirst() {
        `when`(dataStoreFactory
                .cloud(Any::class.java)
                .dynamicGetList<Any>(anyString(), anyString(), anyObject(), anyBoolean(), anyBoolean()))
                .thenReturn(Flowable.just(listOf()))
        `when`(dataStoreFactory
                .disk(Any::class.java)
                .dynamicGetList<Any>(anyString(), anyString(), anyObject(), anyBoolean(), anyBoolean()))
                .thenReturn(Flowable.just(listOf()))
        `when`(dataStoreFactory
                .memory()?.getAllItems(anyObject<Class<Any>>()))
                .thenReturn(Single.just<List<Any>>(listOf(true)))

        dataService.getListOffLineFirst<Any>(getRequest)

        verify<DataStore>(dataStoreFactory.cloud(Any::class.java), times(1))
                .dynamicGetList<Any>(anyString(), anyString(), anyObject(), anyBoolean(), anyBoolean())
        verify<DataStore>(dataStoreFactory.disk(Any::class.java), times(1))
                .dynamicGetList<Any>(anyString(), anyString(), anyObject(), anyBoolean(), anyBoolean())
    }

    @Test
    @Throws(Exception::class)
    fun getObjectOffLineFirst() {
        `when`(dataStoreFactory.cloud(Any::class.java)
                .dynamicGetObject<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(flowable)
        `when`(dataStoreFactory
                .disk(Any::class.java)
                .dynamicGetObject<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(flowable)

        `when`(dataStoreFactory
                .memory()
                ?.getItem<Any>(anyString(), anyObject()))
                .thenReturn(Single.just(true))

        dataService.getObjectOffLineFirst<TestModel>(getRequest)

        verify<DataStore>(dataStoreFactory.cloud(Any::class.java), times(1))
                .dynamicGetObject<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean())
        verify<DataStore>(dataStoreFactory.disk(Any::class.java), times(1))
                .dynamicGetObject<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean())
    }

    @Test
    @Throws(Exception::class)
    fun patchObject() {
        `when`(dataStoreFactory.dynamically(anyString(), anyObject())
                .dynamicPatchObject<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(single)

        dataService.patchObject<TestModel>(postRequest)

        verify<DataStore>(dataStoreFactory.dynamically(anyString(), anyObject()), times(1))
                .dynamicPatchObject<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean())
    }

    @Test
    @Throws(Exception::class)
    fun postObject() {
        `when`(dataStoreFactory
                .dynamically(anyString(), anyObject())
                .dynamicPostObject<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(single)

        dataService.postObject<TestModel>(postRequest)

        verify<DataStore>(dataStoreFactory.dynamically(anyString(), anyObject()), times(1))
                .dynamicPostObject<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean())
    }

    @Test
    @Throws(Exception::class)
    fun postList() {
        `when`(dataStoreFactory
                .dynamically(anyString(), anyObject())
                .dynamicPostList<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(single)

        dataService.postList<TestModel>(postRequest)

        verify<DataStore>(dataStoreFactory.dynamically(anyString(), anyObject()), times(1))
                .dynamicPostList<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean())
    }

    @Test
    @Throws(Exception::class)
    fun putObject() {
        `when`(dataStoreFactory
                .dynamically(anyString(), anyObject())
                .dynamicPutObject<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(single)

        dataService.putObject<TestModel>(postRequest)

        verify<DataStore>(dataStoreFactory.dynamically(anyString(), anyObject()), times(1))
                .dynamicPutObject<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean())
    }

    @Test
    @Throws(Exception::class)
    fun putList() {
        `when`(dataStoreFactory
                .dynamically(anyString(), anyObject())
                .dynamicPutList<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(single)

        dataService.putList<TestModel>(postRequest)

        verify<DataStore>(dataStoreFactory.dynamically(anyString(), anyObject()), times(1))
                .dynamicPutList<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean())
    }

    @Test
    @Throws(Exception::class)
    fun deleteItemById() {
        `when`(dataStoreFactory
                .dynamically(anyString(), anyObject())
                .dynamicDeleteCollection<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(single)

        dataService.deleteItemById<TestModel>(postRequest)

        verify<DataStore>(dataStoreFactory.dynamically(anyString(), anyObject()), times(1))
                .dynamicDeleteCollection<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean())
    }

    @Test
    @Throws(Exception::class)
    fun deleteCollection() {
        `when`(dataStoreFactory
                .dynamically(anyString(), anyObject())
                .dynamicDeleteCollection<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean()))
                .thenReturn(single)

        dataService.deleteCollectionByIds<TestModel>(postRequest)

        verify<DataStore>(dataStoreFactory.dynamically(anyString(), anyObject()), times(1))
                .dynamicDeleteCollection<Any>(
                        anyString(),
                        anyString(),
                        anyObject(),
                        anyObject(),
                        anyObject(),
                        anyObject(),
                        anyBoolean(),
                        anyBoolean())
    }

    @Test
    @Throws(Exception::class)
    fun deleteAll() {
        `when`(dataStoreFactory.disk(Any::class.java).dynamicDeleteAll(anyObject()))
                .thenReturn(Single.just(true))

        dataService.deleteAll(postRequest)

        verify<DataStore>(dataStoreFactory.disk(Any::class.java), times(1)).dynamicDeleteAll(anyObject())
    }

    @Test
    @Throws(Exception::class)
    fun queryDisk() {
        `when`(dataStoreFactory.disk(Any::class.java)
                .queryDisk(anyString(), anyObject<Class<TestModel>>()))
                .thenReturn(Flowable.just(TestModel()))

        dataService.queryDisk("", TestModel::class.java)

        verify<DataStore>(dataStoreFactory.disk(Any::class.java), times(1))
                .queryDisk(anyString(), anyObject<Class<TestModel>>())
    }

    @Test
    @Throws(Exception::class)
    fun uploadFile() {
        `when`<Single<File>>(dataStoreFactory
                .cloud(Any::class.java)
                .dynamicUploadFile(
                        anyString(),
                        anyMap() as HashMap<String, File>,
                        anyMap() as HashMap<String, Any>,
                        anyObject()))
                .thenReturn(fileFlowable)

        dataService.uploadFile<Any>(
                FileIORequest.Builder("", File("")).responseType(Any::class.java).build())

        verify<DataStore>(dataStoreFactory.cloud(Any::class.java), times(1))
                .dynamicUploadFile<Any>(
                        anyString(),
                        anyMap() as HashMap<String, File>,
                        anyMap() as HashMap<String, Any>,
                        anyObject())
    }

    @Test
    @Throws(Exception::class)
    fun downloadFile() {
        `when`<Single<File>>(dataStoreFactory
                .cloud(Any::class.java)
                .dynamicDownloadFile(
                        anyString(),
                        anyObject()))
                .thenReturn(fileFlowable)

        dataService.downloadFile(
                FileIORequest.Builder("", File("")).responseType(Any::class.java).build())

        verify<DataStore>(dataStoreFactory.cloud(Any::class.java), times(1))
                .dynamicDownloadFile(anyString(), anyObject())
    }
}
